[NFC] Add SILBuilderContext.

atrick opened the pull request:
In an upcoming bug fix, I want to pass SILBuilderContext to a
utility. I could continue reusing SILBuilder, even though the utility
must set its own insertion point and debug location. However, this is
terrible practice that I don't want to perpetuate.

The lifetime of a SILBuilder should correspond to a single insertion
point and debug location. That's the only sane way to preserve debug
information in SIL passes.

There are various pieces of contextual state that we've been adding to
the SILBuilder. Those have made it impossible to use SILBuilder
correctly. I'm pulling the context out, so clients can begin using
better practices. In the future, we can make SILBuilderContext
polymorphic, so passes can extend it easily with arbitrary
callbacks. We can also make it self-contained so we don't need to pass
around pointers to an InsertedInst list anymore.

atrick commented:
@swift-ci smoke test.

vedantk commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -31,27 +31,25 @@ class IntegerLiteralExpr;
-class SILBuilder {
-  friend class SILBuilderWithScope;
+/// Manage the state needed for a SIL pass across multiple, independent
+/// SILBuilder invocations.
```
I think it would help to give an explicit usage directive. Something along the lines of: 'use this to construct fresh builder instances instead of repeatedly changing a single builder's insertion point'.

atrick replied:
I added:
```
/// Manage the state needed for a SIL pass across multiple, independent
/// SILBuilder invocations.
///
/// A SIL pass can instantiate a SILBuilderContext object to track information
/// across multiple, potentially independent invocations of SILBuilder. This
/// allows utilities used within the pass to construct a new SILBuilder instance
/// whenever it is convenient or appropriate. For example, a separate SILBuilder
/// should be constructed whenever the current debug location or insertion point
/// changed. Reusing the same SILBuilder and calling setInsertionPoint() easily
/// leads to incorrect debug information.
```
Feel free to add to or rewrite in follow up PRs.

vedantk commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -31,27 +31,25 @@ class IntegerLiteralExpr;
-  SILBasicBlock *BB;
-  SILBasicBlock::iterator InsertPt;
-  const SILDebugScope *CurDebugScope = nullptr;
-  Optional<SILLocation> CurDebugLocOverride = None;
```
The explicit purpose of the debug location override is to survive changes to the insertion point and debug location. It's used in a case where the builder must emit a batch of instructions with the same location (with exceptions for debug intrinsics), and would be cumbersome to do any other way.

Could you preserve the location override in SILBuilderContext?

atrick replied:
I'm baffled by this. A batch of instructions with the same debug location should always use the same SILBuilder instance. The point of a separate SILBuilderContext is that some of the state was totally independent of the debug location or insertion point.

vedantk replied:
My point is that a location override is independent of the builder's debug location or insertion point, which fits your description of state managed by SILBuilderContext. In a model where you need a new builder instance to change insertion point, it may not be true that a batch of instructions which should have the same debug location will be emitted by a single builder.

atrick replied:
We have several broken abstractions built up around SILBuilder. DebugLocOverrideRAII is one of those. Each one of those broken abstractions makes the problem worse in the sense that it makes the responsibility of each type less clear, harder to reason about correctness, and harder to spot bugs. Adding a debug location override to SILBuilderContext would break the abstraction and violate separation of concerns. I need this PR so I can use SILBuilder in utilities as it was originally intended, not so we can have more broken abstractions and make it easier to misuse SILBuilder.

The point is, if you see a SILBuilder being instantiated, you should know what debug location it will use. If a pass is doing something more complicated, it's either going to have it's own wrapper around a SILBuilder, or will need to *explicitly* reuse the builder and *explicitly* modify the debug location or insertion point. In the case of DebugLocOverrideRAII, SILGenBuilder should manage it's own current debug location, it's very easy to do this explicitly without breaking the underlying SILBuilder abstractions.

vedantk replied:
I think limiting usage of the location override to SILGenBuilder would be fine, with the caveat that it would require making getSILDebugLocation virtual. If we can agree that the tradeoff is worthwhile, I have no objections to this patch.

atrick replied:
I can think of many ways to solve SILGen's getSILDebugLocation issue, but it all depends on how code will be migrated, so we should probably discuss that in a different PR. The mere existence of the override mechanism is direct evidence that the API is severely broken and being misused.

vedantk reviewed (commented):
Thanks for working on this.

I agree that creating a fresh builder instance is the cleanest way to change insertion point. Compared to the alternative of passing location and scope information to setInsertionPoint, it avoids mess related to restoring old insertion points. One downside is that it makes for a more involved refactor, but we can make these changes incrementally if needed.

gottesmm commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -31,27 +31,33 @@ class IntegerLiteralExpr;
+  ///
+  /// TODO: Give this ownership of InsertedInstrs and migrate users that
+  /// currently provide their own InsertedInstrs.
+  InsertedInstVec *InsertedInstrs = nullptr;
```
Two questions here:

1. Why use the typedef here and not everywhere else (i.e. constructor)? What value does it provide? I would rather have the real type here given it is just SmallVectorImpl<SILInstruction *>.
2. What are you imagining as the new design? The builder with a TinyPtrVector?

atrick replied:
It looks like I just forgot to remove the useless typedef. I would imagine a plain old SmallVector would be fine, and we could reexpose that to clients if needed as a SmallVectorImpl. InsertedInstrs could be moved into a subclass if we feel like it. I added SILBuilderContext::notifyInserted(SILInstruction*) to make the long term intention more clear.

atrick committed:
[NFC] Add SILBuilderContext.

In an upcoming bug fix, I want to pass SILBuilderContext to a
utility. I could continue reusing SILBuilder, even though the utility
must set its own insertion point and debug location. However, this is
terrible practice that I don't want to perpetuate.

The lifetime of a SILBuilder should correspond to a single insertion
point and debug location. That's the only sane way to preserve debug
information in SIL passes.

There are various pieces of contextual state that we've been adding to
the SILBuilder. Those have made it impossible to use SILBuilder
correctly. I'm pulling the context out, so clients can begin using
better practices. In the future, we can make SILBuilderContext
polymorphic, so passes can extend it easily with arbitrary
callbacks. We can also make it self-contained so we don't need to pass
around pointers to an InsertedInst list anymore.

atrick commented:
@swift-ci smoke test.
